### DISCUSSION TITLE
In what circumstances is context not passed from onMutate to onSuccess?

### DISCUSSION THREAD

**Discussion by RazerM** [TanStack/query#4204] [Q&A]
I apologise if this has previously been discussed, there are a lot of search results for "context" so I may have missed something.

The type of the useQuery callbacks (`onSuccess`/`onError`/`onSettled`) have `context: TContext | undefined`.

If my `onMutate` function returns a type `T`, is that not guaranteed to be passed to the `context` parameter of the callbacks?

**Comment by TkDodo** [comment]
onMutate is async, so it can return a failed / rejected promise, so the return part is never reached if that happens.

**Reply by RazerM** [reply]
Thanks, so that's clear for onError and onSettled then. If there's nothing else, I *think* on the `onMutate` type could be changed from

https://github.com/TanStack/query/blob/5e37b145a8836b00dd20860e8eb1f94c608fec81/src/core/types.ts#L545-L547

to

```ts
  onMutate?: (
    variables: TVariables
  ) => Promise<TContext> | TContext
```

which would allow `onSuccess` to be

```ts
  onSuccess?: (
    data: TData,
    variables: TVariables,
    context: TContext
  ) => Promise<unknown> | void
```

onError and onSettled would maintain `TContext | undefined` obviously.

**Reply by RazerM** [reply]
It would be nice to document this too, that context may only be undefined if onMutate throws/rejects (unless TContext includes undefined of course)

I could contribute this if my proposed changes aren't invalid.

**Reply by TkDodo** [reply]
> I think on the onMutate type could be changed

I think you are right. would you like to try this in a PR ?
